provers/groth16/src/setup.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{common::*, QAP};
+use lambdaworks_math::{
+    cyclic_group::IsGroup,
+    elliptic_curve::traits::{IsEllipticCurve, IsPairing},
+};
+
+pub struct VerifyingKey {
+    // e([alpha]_1, [beta]_2) computed during setup as it's a constant
+    pub alpha_g1_times_beta_g2: PairingOutput,
+    pub delta_g2: G2Point,
+    pub gamma_g2: G2Point,
+    // [K_0(τ)]_1, [K_1(τ)]_1, ..., [K_k(τ)]_1
+    // where K_i(τ) = γ^{-1} * (β*l(τ) + α*r(τ) + o(τ))
+    // and "k" is the number of public inputs
+    pub verifier_k_tau_g1: Vec<G1Point>,
+}
+
+pub struct ProvingKey {
+    pub alpha_g1: G1Point,
+    pub beta_g1: G1Point,
+    pub beta_g2: G2Point,
+    pub delta_g1: G1Point,
+    pub delta_g2: G2Point,
+    // [A_0(τ)]_1, [A_1(τ)]_1, ..., [A_n(τ)]_1
+    pub l_tau_g1: Vec<G1Point>,
+    // [B_0(τ)]_1, [B_1(τ)]_1, ..., [B_n(τ)]_1
+    pub r_tau_g1: Vec<G1Point>,
+    // [B_0(τ)]_2, [B_1(τ)]_2, ..., [B_n(τ)]_2
+    pub r_tau_g2: Vec<G2Point>,
+    // [K_{k+1}(τ)]_1, [K_{k+2}(τ)]_1, ..., [K_n(τ)]_1
+    // where K_i(τ) = ƍ^{-1} * (β*l(τ) + α*r(τ) + o(τ))
+    // and "k" is the number of public inputs
+    pub prover_k_tau_g1: Vec<G1Point>,
+    // [delta^{-1} * t(τ) * tau^0]_1, [delta^{-1} * t(τ) * τ^1]_1, ..., [delta^{-1} * t(τ) * τ^m]_1
+    pub z_powers_of_tau_g1: Vec<G1Point>,
+}
+
+struct ToxicWaste {
+    tau: FrElement,
+    alpha: FrElement,
+    beta: FrElement,
+    gamma: FrElement,
+    delta: FrElement,
+}
+
+impl ToxicWaste {
+    pub fn new() -> Self {
+        Self {
+            tau: sample_fr_elem(),
+            alpha: sample_fr_elem(),
+            beta: sample_fr_elem(),
+            gamma: sample_fr_elem(),
+            delta: sample_fr_elem(),
+        }
+    }
+}
+
+pub fn setup(qap: &QAP) -> (ProvingKey, VerifyingKey) {
+    let g1: G1Point = Curve::generator();
+    let g2: G2Point = TwistedCurve::generator();
+
+    let tw = ToxicWaste::new();
+
+    let num_of_total_inputs = qap.l.len();
+
+    // [A_i(τ)]_1, [B_i(τ)]_1, [B_i(τ)]_2, [K_i(τ)]_1
+    let mut l_tau_g1: Vec<G1Point> = Vec::with_capacity(num_of_total_inputs);
+    let mut r_tau_g1: Vec<G1Point> = Vec::with_capacity(num_of_total_inputs);
+    let mut r_tau_g2: Vec<G2Point> = Vec::with_capacity(num_of_total_inputs);
+    let mut verifier_k_tau_g1: Vec<G1Point> = Vec::with_capacity(qap.num_of_public_inputs);
+    let num_of_private_inputs = num_of_total_inputs - qap.num_of_public_inputs;
+    let mut prover_k_tau_g1: Vec<G1Point> = Vec::with_capacity(num_of_private_inputs);
+
+    let delta_inv = tw.delta.inv().unwrap();
+    let gamma_inv = tw.gamma.inv().unwrap();
+    for i in 0..num_of_total_inputs {
+        let l_i_tau = qap.l[i].evaluate(&tw.tau);
+        let r_i_tau = qap.r[i].evaluate(&tw.tau);
+
+        l_tau_g1.push(g1.operate_with_self(l_i_tau.representative()));
```

Comment:
I think we could make some use of successors or other iterators, mainly to avoid creating the vector directly and then push

Reply:
I switched to a more functional style of code with iterators